### PR TITLE
Elwood Adapter

### PR DESCRIPTION
**Pull request by austinborn**
## Description
Build new Elwood external adapter for WS-only connections

## Changes
- Add Elwood EA for WS price updates only
- Add `await` for WS subscription (since Elwood subscriptions are through REST)
- Remove `console.log` from WS test framework

## Steps to Test
`yarn && yarn setup && yarn test packages/sources/elwood/test/`

## Quality Assurance
- [x] If a new adapter was made, or an existing one was modified so that its environment variables have changed, update the relevant `<ADAPTER_PACKAGE>/schemas/env.json` and `<ADAPTER_PACKAGE>/README.md`
- [x] If a new adapter was made, or an existing one was modified so that its environment variables have changed, update the relevant `infra-k8s` configuration file.
- [x] The branch naming follows git flow (`feature/x`, `chore/x`, `release/x`, `hotfix/x`, `fix/x`) or is created from Clubhouse/Shortcut
- [x] This is related to a maximum of one Clubhouse/Shortcut story or GitHub issue
- [ ] Types are safe (avoid TypeScript/TSLint features like any and disable, instead use more specific types)
- [x] All code changes have 100% unit and integration test coverage. If testing is not applicable or too difficult to justify doing, the reasoning should be documented explicitly in the PR.

### CODE REVIEW

**Comment by justinkaseman** [comment]
Try running `yarn` locally to rebuild the lock file

File: packages/sources/elwood/src/adapter.ts
Comment (boxhock) [inline packages/sources/elwood/src/adapter.ts:58]: Can we get this TODO resolved before merging?

File: packages/sources/elwood/src/adapter.ts
Comment (boxhock) [inline packages/sources/elwood/src/adapter.ts:79]: Can you add some clarification as to what's happening here? It seems this is a WS handler, but we're also making a POST request without catching the response data?

File: packages/sources/elwood/src/adapter.ts
Comment (boxhock) [inline packages/sources/elwood/src/adapter.ts:83]: Can we avoid `any`?

File: packages/sources/elwood/src/adapter.ts
Comment (boxhock) [inline packages/sources/elwood/src/adapter.ts:95]: Is this logging needed?

File: packages/sources/elwood/src/adapter.ts
Comment (boxhock) [inline packages/sources/elwood/src/adapter.ts:102]: Is this logging needed?

**Review by boxhock — CHANGES_REQUESTED** [review]

File: packages/sources/elwood/src/adapter.ts
Comment (austinborn) [reply packages/sources/elwood/src/adapter.ts:102]: At this point I'll remove it

File: packages/sources/elwood/src/adapter.ts
Comment (austinborn) [reply packages/sources/elwood/src/adapter.ts:95]: At this point I'll remove it

File: packages/sources/elwood/src/adapter.ts
Comment (austinborn) [reply packages/sources/elwood/src/adapter.ts:83]: I'll change this to `Error`

File: packages/sources/elwood/src/adapter.ts
Comment (austinborn) [reply packages/sources/elwood/src/adapter.ts:79]: The Elwood API only allows using POST requests to subscribe to WS feeds, so this was my attempt to send the subscription request and return `subscriptionMsg` without needing to mess with the framework. The `Requester.request` call is async, so waiting for the return value would require `awaiting` in every instance of `wsHandler.subscribe` and `wsHandler.unsubscribe`. I'm aware it's not ideal, so we can investigate a longer-term solution if desired.

Comment (boxhock) [reply packages/sources/elwood/src/adapter.ts:79]: Could you add this explanation to the code, so we don't lose this knowledge? :sweat_smile:

File: packages/sources/elwood/src/adapter.ts
Comment (austinborn) [reply packages/sources/elwood/src/adapter.ts]: Removed

File: packages/sources/elwood/src/adapter.ts
Comment (austinborn) [reply packages/sources/elwood/src/adapter.ts]: Removed

File: packages/sources/elwood/src/adapter.ts
Comment (austinborn) [reply packages/sources/elwood/src/adapter.ts]: Updated to `const error = e as Error`

File: packages/sources/elwood/src/adapter.ts
Comment (austinborn) [reply packages/sources/elwood/src/adapter.ts]: Yeah for sure

Comment (austinborn) [reply packages/sources/elwood/src/adapter.ts]: Done

**Review by boxhock — DISMISSED** [review]
Looks good to me, but would like someone to give a second opinion on including `any`here.

**Comment by austinborn** [comment]
> Looks good to me, but would like someone to give a second opinion on including `any`here.

@justinkaseman Any thoughts?

File: packages/sources/elwood/src/adapter.ts
Comment (alejoberardino) [reply packages/sources/elwood/src/adapter.ts:58]: If the response types that are used in the adapter code are all included here, I don't think we need to go and get every possible ws message for the API, as long as the code itself is typesafe I'm good

**Comment by KuphJr** [comment]
> Looks good to me, but would like someone to give a second opinion on including `any`here.

If we could use the `unknown` type and cast it if you try to access a specific property, I think that is the ideal solution.  I didn't spend too much time looking at all the code though, so if that is also going to be too difficult, I think using `any` is fine.  Just be sure to check that all the properties you access are not `undefined` so you don't cause a crash.

**Comment by austinborn** [comment]
> > Looks good to me, but would like someone to give a second opinion on including `any`here.
> 
> If we could use the `unknown` type and cast it if you try to access a specific property, I think that is the ideal solution. I didn't spend too much time looking at all the code though, so if that is also going to be too difficult, I think using `any` is fine. Just be sure to check that all the properties you access are not `undefined` so you don't cause a crash.

I get the following when I try to set it to `unknown`:
```
packages/sources/elwood/src/index.ts:11:58 - error TS2345: Argument of type 'MakeWSHandler<unknown>' is not assignable to parameter of type 'MakeWSHandler<UnknownWSMessage>'.
  Type 'WSHandler<unknown> | Promise<WSHandler<unknown>>' is not assignable to type 'WSHandler<UnknownWSMessage> | Promise<WSHandler<UnknownWSMessage>>'.
    Type 'WSHandler<unknown>' is not assignable to type 'WSHandler<UnknownWSMessage> | Promise<WSHandler<UnknownWSMessage>>'.
      Type 'WSHandler<unknown>' is not assignable to type 'WSHandler<UnknownWSMessage>'.
        The types returned by 'subscribe(...)' are incompatible between these types.
          Type 'unknown' is not assignable to type 'UnknownWSMessage | undefined'.

11 const { server } = expose(adapterContext, makeExecute(), makeWSHandler(), endpointSelector)
```

File: packages/sources/elwood/src/adapter.ts
Comment (austinborn) [reply packages/sources/elwood/src/adapter.ts:58]: I can add `util.isObject` checks to ensure we at least know the input is an object before trying to operate on it

**Comment by boxhock** [comment]
@austinborn what about `UnknownWSMessage`?

**Review by boxhock — APPROVED** [review]

**Pull request merged**
